File: crates/defi-cli/src/commands/schema.rs
use crate::output::OutputMode;
use clap::Args;
use defi_core::error::Result;

#[derive(Args)]
pub struct SchemaArgs {
    /// Command to get schema for (e.g. "dex.swap", "lending.supply")
    pub command: Option<String>,
    /// Show all schemas
    #[arg(long)]
    pub all: bool,
}

pub async fn run(args: SchemaArgs, output: &OutputMode) -> Result<()> {
    let action = if args.all {
        "all".to_string()
    } else {
        args.command.unwrap_or_else(|| "all".to_string())
    };
    let params = serde_json::json!({"action": action});
    let schema = crate::agent::handle_schema(&params)?;
    output.print(&schema)?;
    Ok(())
}

File: ts/packages/defi-cli/src/commands/schema.ts
import type { Command } from "commander";
import type { OutputMode } from "../output.js";
import { printOutput } from "../output.js";
import { handleSchema } from "../agent.js";

export function registerSchema(parent: Command, getOpts: () => OutputMode): void {
  parent
    .command("schema [command]")
    .description("Output JSON schema for a command (agent-friendly)")
    .option("--all", "Show all schemas")
    .action(async (command: string | undefined, opts: { all?: boolean }) => {
      const mode = getOpts();
      // CLI passes hyphenated forms (`schema lending-supply`); the action
      // dispatcher in agent.ts uses dotted keys (`lending.supply`). Normalise
      // here so the schema lookup hits the right case instead of falling
      // through to the generic action list.
      const raw = opts.all ? "all" : (command ?? "all");
      const action = raw.replace(/-/g, ".");
      const params: Record<string, unknown> = { action };
      const schema = handleSchema(params);
      printOutput(schema, mode);
    });
}
